pyccel/stdlib/lists/lists.c

```diff
@@ -0,0 +1,665 @@
+ * every value.
+ * Returns comparision value.
+ * 
+ * @param list 
+ * @param i1 
+ * @param i2 
+ * @return int 
+ */
+int compare(t_list *list, int i1, int i2)
+{
+    int8_t *group_1;
+    int8_t *group_2;
+    int8_t type;
+    size_t len;
+    int cmp;
+
+    cmp = 0;
+    len = 0;
+    type = 0;
+    if (list->type != lst_list)
+    {
+        switch (list->type)
+        {
+            case lst_int8:
+                return *((int8_t *)GET_INDEX(list, i1)) - *((int8_t *)GET_INDEX(list, i2));
+            case lst_int16:
+                return *((int16_t *)GET_INDEX(list, i1)) - *((int16_t *)GET_INDEX(list, i2));
+            case lst_int32:
+                return *((int32_t *)GET_INDEX(list, i1)) - *((int32_t *)GET_INDEX(list, i2));
+            case lst_int64:
+                return *((int64_t *)GET_INDEX(list, i1)) - *((int64_t *)GET_INDEX(list, i2));
+            case lst_float:
+                return *((float *)GET_INDEX(list, i1)) - *((float *)GET_INDEX(list, i2));
+            case lst_double:
+                return *((double *)GET_INDEX(list, i1)) - *((double *)GET_INDEX(list, i2));
+            case lst_complex:
+                fprintf(stderr, "Ordering of complex numbers is not supported.");
+                exit(-1);
+        }
+    }
+
+    set_length_and_type(((t_list **)list->elements)[0], &len, &type);
+    group_1 = group_node_items(((t_list **)list->elements)[i1], len, type);
+    group_2 = group_node_items(((t_list **)list->elements)[i2], len, type);
+
+    for (size_t i = 0; i < len && cmp == 0; i++)
+    {
+        switch (type)
+        {
+            case lst_int8:
+                cmp = ((int8_t *)group_1)[i] - ((int8_t *)group_2)[i];
+                break ;
+            case lst_int16:
+                cmp = ((int16_t *)group_1)[i] - ((int16_t *)group_2)[i];
+                break ;
+            case lst_int32:
+                cmp = ((int32_t *)group_1)[i] - ((int32_t *)group_2)[i];
+                break ;
+            case lst_int64:
+                cmp = ((int64_t *)group_1)[i] - ((int64_t *)group_2)[i];
+                break ;
+            case lst_float:
+                cmp = ((float *)group_1)[i] - ((float *)group_2)[i];
+                break ;
+            case lst_double:
+                cmp = ((double *)group_1)[i] - ((double *)group_2)[i];
+                break ;
+            case lst_complex:
+                fprintf(stderr, "Ordering of complex numbers is not supported.");
+                exit(-1);
+        }
+    }
+
+    free(group_1);
+    free(group_2);
+    return cmp;
+}
+
+/**
+ * @brief Always take the last element of the list and use it as an anchor
+ * for sorting the elements with values less than it's value, while updating 
+ * the index where the anchor should exist, then swap the value in the index 
+ * `r` (the current position of the anchor) with new found position, this 
+ * ensures that the anchor is always in the right position.
+ * Returns the index of the anchor.
```

Comment:
This docstring describes the method used, but it misses a description of what you are trying to do. What is `partition`?

Reply:
This function will not remain in the code as I opted to replace it with the c function `qsort_r(3)`, which I think it uses the same quicksort algorithm, but a lot faster.

Reply:
https://github.com/pyccel/pyccel/pull/1192/commits/e7d7fce524d55ba8528fe125504613a3bdb7fb27